### ISSUE TITLE
\newcommand redefines TeX commands (such as \begin, \mathbf, ...)

### ISSUE THREAD

**Issue by SpectralFlame** [mathjax/MathJax#2934]
### Issue Summary

When defining a new command (using `\newcommand` or `\def`) on a line **before** a command such as `\begin{...}`; in the process of typing that, it will redefine `\begin` to whatever is in the braces, essentially breaking any environments defined with `\begin`.
This issue is not specific to `\begin`, though.

### Steps to Reproduce:

1. Starting point
```
$$
\begin{aligned}
a = 3
\end{aligned}
$$
```
2. Type `\newcommand` on the first line to get
```
$$
\newcommand

\begin{aligned}
a = 3
\end{aligned}
$$
```

At this point, `\begin` is redefined to `aligned`, even though `\begin` is already a defined command. Also, the purpose was to define another command entirely, but in the process of writing it, something else got redefined too.
Ideally, one would be able to define a **new** command at the top without having to worry that stuff could break by accident while typing out the command.

### Technical details:

* MathJax Version: 3.2
* Client OS: Linux 5.19.9-arch1-1
* Browser: Vivaldi 5.4.2753.51 (Stable channel) stable (64-bit) 

I am using the following MathJax configuration:
Live demo on [https://www.mathjax.org/#demo](https://www.mathjax.org/#demo)

**Comment by dpvc** [comment]
I'm not sure what you are asking to be done.  In LaTeX, there are two separate commands, `\newcommand` and `\renewcommand`, where the first will produce an error if you try to redefine an existing command, while the second produces an error if you use it for a command that *doesn't* already exist.  It is true that in MathJax, neither of those checks is performed (and they are both essentially the same).  

Are you asking to have the same error conditions checked as in LaTeX?  That will make the `\newcommand` case work for you, but will still be problematic for if you re trying to type a `\renewcommand` before other macros.  I do not see a way around that.  Do you have a suggestion on how it should be handled?  I think it is something you will just have to be aware of when you type in a setting where what you type is being reprocessed as you type.  One could, for example, put the `\newcommand` in separate math delimiters to avoid the issue, or enclose it in braces until you are done typing.

**Comment by SpectralFlame** [comment]
After giving it some thought, I think a pretty straightforward way to avoid this issue would be to insert `{}{}` after a newly typed `\newcommand`, `\renewcommand` or `\def` is detected. In the case of pasting, if it (ignoring whitespaces) ends in one of the aforementioned commands, it should probably also insert `{}{}` to be safe.

I'm not sure if the cursor should be placed within the first `{}` or not, since users might just keep typing without noticing they are inside the first argument, but I think that is still better than potentially breaking some macro and needing to restart the application to fix it.

**Comment by dpvc** [comment]
What you suggest would certainly work, but note that it is a solution based on the *editor* being used, not on MathJax itself.  The demo page is really just a simple textarea element that copies its content to the display area and calls MathJax to typeset it after each keypress.  It is not a fancy editor, and is only meant as a simple playground, not a sophisticated text editor.  While we could add more features to it, like the braces you are suggesting, that would only fix the issue on *our* page, not for any other page that allows you to enter LaTeX notation (e.g., the StackExchange sites).

So this is not an issue that is solved within MathJax itself, but within the framework that is *using* MathJax.  I will leave this open as a feature request for the demo page, but it is not likely to see any action in the near future.

**Comment by SpectralFlame** [comment]
I thought this might be the case.

The only other option I could think of, and which could be implemented in MathJax, would be to temporarily stop reprocessing when a `\newcommand`, `\renewcommand` or `\def` is detected, until
- `\` is typed; e.g. `\newcommand\`
- `{` is typed; e.g. `\newcommand{`
- the cursor has moved
This will raise an error rather than prematurely redefining a macro, until the (re)new command is fully defined.

**Comment by hubgit** [comment]
I've observed a possibly-related issue, which is that MathJax seems to remember everything that's been defined with `\newcommand`, even if it the definition isn't present any more.

For example, enter this into <https://www.mathjax.org/#demo>:

```
$\newcommand{\foo}{\alpha}$

$\foo$
```

then change the definition:

```
$\newcommand{\bar}{\alpha}$

$\foo$
```

Observe that the `α` is still rendered, even though `\foo` is no longer defined.

If you typed `\bar` manually, one character at a time, this also means that `\b` and `\ba` are now defined.

**Comment by dpvc** [comment]
@hugbit, yes, definitions are permanent to the page where they are made, so if you make a definition, it will stay in effect until you reload the page (regardless of whether it is currently in the demo textarea.  This is because, if you are using MathJax on a page with dynamic content, and you made a definition someone on the page, then add more content that uses that definition, you would want that to still be available.  The content you are suing is being added to the page in that way, so any previous definitions are still in place.  As I mentioned above, the demo is just a simple playground, not a fancy editor.

**Comment by hubgit** [comment]
@dpvc It's actually in the context of a dynamic editor that I'm observing this issue, where equations that use a custom command continue to render without error even after the custom command definition is no longer in the page.

Might there be a way to reset the cached commands, like there is for resetting the equation numbering, or a configuration setting to disable this behaviour?

**Comment by dpvc** [comment]
@hugbit, the `begingroup` extension from v2 is one of the last remaining ones to be ported over to v3 and above.  That is the one what localizes definitions, and could be used to do what you are looking for.  But for now, that functionality is not yet available.

**Comment by dpvc** [comment]
Closing this as `begingroup` is available in v4.0, and no further work is expected to be done on this issue.
